### PR TITLE
Moved on from lazy shared `HttpClient` initialization using a null check to using an enum in `HttpClientProvider` implementations

### PR DESCRIPTION
**Pull request by vcolin7**
Fixes: #39766.

Also updated tests.

### CODE REVIEW

File: sdk/clientcore/http-jdk-httpclient/src/test/java/io/clientcore/http/jdk/httpclient/JdkHttpClientSingletonTests.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import io.clientcore.core.http.client.HttpClient;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.parallel.Execution;
+import org.junit.jupiter.api.parallel.ExecutionMode;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+@Execution(ExecutionMode.SAME_THREAD) // To avoid a race condition because of singleton http client.
```
Comment (srnagar) [inline sdk/clientcore/http-jdk-httpclient/src/test/java/io/clientcore/http/jdk/httpclient/JdkHttpClientSingletonTests.java:14]: Why is this needed? Since we expect customer apps also to be multithreaded, these tests should be able to handle multithreading.

**Review by srnagar — APPROVED** [review]

File: sdk/clientcore/http-jdk-httpclient/src/test/java/io/clientcore/http/jdk/httpclient/JdkHttpClientSingletonTests.java (hunk 1 above)
Comment (vcolin7) [reply sdk/clientcore/http-jdk-httpclient/src/test/java/io/clientcore/http/jdk/httpclient/JdkHttpClientSingletonTests.java:14]: I copied the tests over from Azure Core. I can remove it and see how the tests behave.

File: sdk/clientcore/core/src/main/java/io/clientcore/core/http/client/HttpClientProvider.java [hunk 2]
```diff
@@ -32,13 +45,7 @@ public abstract class HttpClientProvider {
...
      * create.
      */
-    public final HttpClient getSharedInstance() {
-        if (sharedHttpClient == null) {
-            sharedHttpClient = getNewInstance();
-        }
-
-        return sharedHttpClient;
-    }
+    public abstract HttpClient getSharedInstance();
```
Comment (JonathanGiles) [inline sdk/clientcore/core/src/main/java/io/clientcore/core/http/client/HttpClientProvider.java:48]: Should we implement this method here with the default implementation? That way presumably `GlobalDefaultHttpClient` can be made private as well

Comment (vcolin7) [reply sdk/clientcore/core/src/main/java/io/clientcore/core/http/client/HttpClientProvider.java:48]: The reason `GlobalDefaultHttpClient` is package-private is so it can be used in `HttpClient.getSharedInstance()`:
```java
static HttpClient getSharedInstance() {
    return HttpClientProvider.getProviders().create(HttpClientProvider::getSharedInstance,
        HttpClientProvider.GlobalDefaultHttpClient.HTTP_CLIENT::getHttpClient, null);
}
```

As the `fallbackSupplier` in the `Providers.create()` method:

```java
public TInstance create(Function<TProvider, TInstance> createInstance, Supplier<TInstance> fallbackSupplier,
    Class<? extends TProvider> selectedImplementation)
```

It's possible I don't understand the `Providers` logic entirely, but it seems to me that if we want to keep the `HttpClient.getSharedInstance()` method functionality, `GlobalDefaultHttpClient` will still need to be exposed to `HttpClient` somehow to provide a `fallbackSupplier`.

**Review by alzimmermsft — APPROVED** [review]
